src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/MissingWorkloadRegistrationService.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            if (!WorkloadDescriptor.Empty.Equals(workloadDescriptor))
+            {
+                if (!_projectGuidToWorkloadDescriptorsMap.TryGetValue(projectGuid, out var workloadDescriptorSet))
+                {
+                    workloadDescriptorSet = new HashSet<WorkloadDescriptor>();
+                    _projectGuidToWorkloadDescriptorsMap[projectGuid] = workloadDescriptorSet;
+                }
+
+                workloadDescriptorSet.Add(workloadDescriptor);
+            }
+
+            if (_projectGuidToProjectConfigurationsMap.TryGetValue(projectGuid, out var projectConfigurationSet))
+            {
+                projectConfigurationSet.Remove(projectConfiguration);
+                if (projectConfigurationSet.Count == 0 && ShouldDisplayMissingComponentsPrompt())
+                {
+                    return DisplayMissingComponentsPromptAsync(cancellationToken);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void RegisterProjectConfiguration(Guid projectGuid, ProjectConfiguration projectConfiguration)
+        {
+            if (!_projectGuidToProjectConfigurationsMap.TryGetValue(projectGuid, out var projectConfigurationSet))
+            {
+                projectConfigurationSet = new HashSet<ProjectConfiguration>();
+                _projectGuidToProjectConfigurationsMap[projectGuid] = projectConfigurationSet;
+            }
+
+            projectConfigurationSet.Add(projectConfiguration);
+        }
+
+        private bool ShouldDisplayMissingComponentsPrompt()
+        {
+            return _projectGuidToProjectConfigurationsMap.Values.All(projectConfigurationSet => projectConfigurationSet.Count == 0);
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            ClearMissingWorkloadMetadata();
+
+            return Task.CompletedTask;
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        public async Task DisplayMissingComponentsPromptAsync(CancellationToken cancellationToken)
+        {
+            if (_projectGuidToWorkloadDescriptorsMap.Count == 0)
+            {
+                return;
+            }
+
+            var serviceBrokerContainer = await _serviceBrokerContainer.GetValueAsync();
+            IServiceBroker? serviceBroker = serviceBrokerContainer?.GetFullAccessServiceBroker();
+            if (serviceBroker == null)
+            {
+                return;
+            }
+
+            var missingWorkloadRegistrationService = await serviceBroker.GetProxyAsync<IMissingComponentRegistrationService>(
+                serviceDescriptor: VisualStudioServices.VS2022.MissingComponentRegistrationService);
+
+            using (missingWorkloadRegistrationService as IDisposable)
+            {
+                if (missingWorkloadRegistrationService != null)
```

Comment:
Little confused by this. According to this, you shouldn't really create the variable outside of the using statement scope: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement#remarks

I don't think the `as Disposable` does anything... since you're making a temporary boxing of the variable that is then not stored... so I don't know if it is actually being disposed? And checking if it is non-null inside the using confuses me even more. If you check if it is null outside the using, then there's no reason to have a using? You can't dispose something that is null. Again, brain confusion. 😵

Reply:
> I don't think the `as Disposable` does anything... since you're making a temporary boxing of the variable that is then not stored... so I don't know if it is actually being disposed? And checking if it is non-null inside the using confuses me even more. If you check if it is null outside the using, then there's no reason to have a using? You can't dispose something that is null. Again, brain confusion. 😵

@MiYanni The service does not derive from IDisposable. See the analyzer notes at https://dev.azure.com/devdiv/DevDiv/_git/DevCore?path=%2Fdoc%2Fanalyzers%2FISB001.md&_a=preview